vignettes/make_vocab_and_lookup.R

```diff
@@ -0,0 +1,201 @@
+# Creates a morphemepiece vocabulary and lookup table, given a processed 
+# wiktionary dump and an existing wordpiece vocabulary.
+# Optionally, a word frequency table from some corpus can be given.
+#
+
+make_vocab_and_lookup <- function(processed_wiktionary, 
+                                  wp_vocab, 
+                                  target_vocab_size = 25000,
+                                  word_freq_tab = NULL) {
+  unnested_processed_wiktionary <- unnest_pw(processed_wiktionary, clean = TRUE)
+  token_counts <- get_token_counts(unnested_processed_wiktionary, word_freq_tab)
+  
+  # punctuation, numbers, [CLS], etc. from wordpiece vocab
+  fancy_wp_tokens <- get_fancy_wp_tokens(wp_vocab)
+
+  actual_wp_words <- get_actual_wp_words(processed_wiktionary, wp_vocab)
+  # short actual words from wp will get a VIP pass directly to vocab.
+  # Though most of them wouldn't be considered words by most people. :)
+  short_words <- actual_wp_words[nchar(actual_wp_words) <= 3]
+  rest_of_words <- actual_wp_words[nchar(actual_wp_words) > 3]
+  # make sure we keep enough tokens to cover the rest of the wp words
+  processed_tokens <- unnested_processed_wiktionary %>% 
+    dplyr::filter(word %in% rest_of_words) %>% 
+    dplyr::distinct(token) %>% 
+    dplyr::pull(token)
+  
+  # we *didn't* clean the processed tokens before this, so clean them now.
+  # This mainly filters out a few capitalized proper nouns that sneaked back in,
+  # but we'll recover the lowercase version of those later.
+  processed_tokens <- processed_tokens[stringr::str_detect(processed_tokens,
+                                                           pattern = "[^a-z#]", 
+                                                           negate = TRUE)]
+  
+  wp_proper_nouns <- get_wp_proper_nouns(processed_wiktionary, wp_vocab)
+  # There's a lot of overlap with "actual" words (e.g. apple). This 
+  # *shouldn't* make a difference, but just to be sure, filter those out:
+  wp_proper_nouns <- setdiff(wp_proper_nouns, actual_wp_words)
+
+  # still need to add in letters, letter pairs for completeness.
+  letter_pairs <- paste0("##", as.vector(outer(letters, letters, FUN = paste0)))
+  # remove most pairs that end in "s"
+  letter_pairs <- letter_pairs[!grepl("##[^sc]s", letter_pairs)]
```

Comment:
This is the (one?) point in the process that was clearly informed by English patterns. Removing most `##*s` tokens makes it more likely that an unknown plural will have the final `##s` token correctly broken off.

Reply:
But what about, say, "loraxes". That'd have to tokenize as "loraxe ##s" rather than "lorax ##es", right? I mean, technically the first part would be other tokenization in either case, but it feels like maybe ##es is useful? It's weird, because our lookup just turns ##es into ##s, so maybe not? Hmm.

Ok, I think I'm on board with your plan. Let's keep that well-loaded "##s" as the piece that comes off, even if TECHNICALLY it ends in "es". That makes sense to me :)